File: rust/src/benchmarks/noise.rs
use super::super::{helper, Benchmark};
use crate::config_i64;

const SYM: [char; 6] = [' ', '░', '▒', '▓', '█', '█'];

#[derive(Clone, Copy)]
struct Vec2 {
    x: f64,
    y: f64,
}

impl Vec2 {
    fn new(x: f64, y: f64) -> Self {
        Self { x, y }
    }
}

struct Noise2DContext {
    rgradients: Vec<Vec2>,
    permutations: Vec<i32>,
    size_val: i64,
}

impl Noise2DContext {
    fn new(size: i64) -> Self {
        let size = size as usize;
        let mut rgradients = vec![Vec2::new(0.0, 0.0); size];
        let mut permutations = vec![0i32; size];

        for i in 0..size {
            let v = helper::next_float(1.0) * std::f64::consts::PI * 2.0;
            rgradients[i] = Vec2::new(v.cos(), v.sin());
            permutations[i] = i as i32;
        }

        for _ in 0..size {
            let a = helper::next_int(size as i32) as usize;
            let b = helper::next_int(size as i32) as usize;
            permutations.swap(a, b);
        }

        Self {
            rgradients,
            permutations,
            size_val: size as i64,
        }
    }

    fn get_gradient(&self, x: i32, y: i32) -> Vec2 {
        let idx = self.permutations[(x as usize) & (self.size_val as usize - 1)]
            + self.permutations[(y as usize) & (self.size_val as usize - 1)];
        self.rgradients[(idx as usize) & (self.size_val as usize - 1)]
    }

    fn get(&self, x: f64, y: f64) -> f64 {
        let x0f = x.floor();
        let y0f = y.floor();
        let x0 = x0f as i32;
        let y0 = y0f as i32;
        let x1 = x0 + 1;
        let y1 = y0 + 1;

        let gradients = [
            self.get_gradient(x0, y0),
            self.get_gradient(x1, y0),
            self.get_gradient(x0, y1),
            self.get_gradient(x1, y1),
        ];

        let origins = [
            Vec2::new(x0f, y0f),
            Vec2::new(x0f + 1.0, y0f),
            Vec2::new(x0f, y0f + 1.0),
            Vec2::new(x0f + 1.0, y0f + 1.0),
        ];

        let p = Vec2::new(x, y);

        let gradient = |orig: Vec2, grad: Vec2, p: Vec2| -> f64 {
            let sp = Vec2::new(p.x - orig.x, p.y - orig.y);
            grad.x * sp.x + grad.y * sp.y
        };

        let v0 = gradient(origins[0], gradients[0], p);
        let v1 = gradient(origins[1], gradients[1], p);
        let v2 = gradient(origins[2], gradients[2], p);
        let v3 = gradient(origins[3], gradients[3], p);

        let lerp = |a: f64, b: f64, v: f64| a * (1.0 - v) + b * v;
        let smooth = |v: f64| v * v * (3.0 - 2.0 * v);

        let fx = smooth(x - origins[0].x);
        let vx0 = lerp(v0, v1, fx);
        let vx1 = lerp(v2, v3, fx);

        let fy = smooth(y - origins[0].y);
        lerp(vx0, vx1, fy)
    }
}

pub struct Noise {
    size_val: i64,
    result_val: u32,
    n2d: Noise2DContext,
}

impl Noise {
    pub fn new() -> Self {
        let size_val = config_i64("Noise", "size");
        let n2d = Noise2DContext::new(size_val);

        Self {
            size_val,
            result_val: 0,
            n2d,
        }
    }
}

impl Benchmark for Noise {
    fn name(&self) -> String {
        "Noise".to_string()
    }

    fn run(&mut self, iteration_id: i64) {
        for y in 0..self.size_val {
            for x in 0..self.size_val {
                let v = self
                    .n2d
                    .get(x as f64 * 0.1, (y + (iteration_id * 128)) as f64 * 0.1)
                    * 0.5
                    + 0.5;
                let idx = (v / 0.2) as usize;
                let idx = if idx < SYM.len() { idx } else { SYM.len() - 1 };
                self.result_val = self.result_val.wrapping_add(SYM[idx] as u32);
            }
        }
    }

    fn checksum(&self) -> u32 {
        self.result_val
    }
}

File: julia/src/benchmarks/Noise.jl
using StaticArrays

mutable struct Noise <: AbstractBenchmark
    size::Int64
    result::UInt32
    rgradients::Vector{SVector{2,Float64}}
    permutations::Vector{Int32}

    function Noise()
        size_val = Helper.config_i64("Noise", "size")

        rgradients = Vector{SVector{2,Float64}}(undef, size_val)
        for i = 1:size_val
            v = Helper.next_float() * 2π
            rgradients[i] = SVector(cos(v), sin(v))
        end

        permutations = collect(Int32(0):Int32(size_val-1))
        for _ = 1:size_val
            a = Helper.next_int(Int32(size_val))
            b = Helper.next_int(Int32(size_val))
            permutations[a+1], permutations[b+1] = permutations[b+1], permutations[a+1]
        end

        new(size_val, UInt32(0), rgradients, permutations)
    end
end

name(b::Noise)::String = "Noise"

@inline @fastmath function lerp(a::Float64, b::Float64, v::Float64)::Float64
    return a + v * (b - a)
end

@inline @fastmath function smooth(v::Float64)::Float64
    v2 = v * v
    return v2 * (3.0 - 2.0 * v)
end

@inline function get_gradient(b::Noise, x::Int32, y::Int32)::SVector{2,Float64}

    idx = b.permutations[(x&(b.size-1))+1] + b.permutations[(y&(b.size-1))+1]
    return b.rgradients[(idx&(b.size-1))+1]
end

@inline @fastmath function get(b::Noise, x::Float64, y::Float64)::Float64
    x0f = floor(x)
    y0f = floor(y)
    x0 = Int32(x0f)
    y0 = Int32(y0f)

    size_mask = b.size - 1

    px0 = (x0 & size_mask) + 1
    px1 = ((x0 + Int32(1)) & size_mask) + 1
    py0 = (y0 & size_mask) + 1
    py1 = ((y0 + Int32(1)) & size_mask) + 1

    perm = b.permutations
    grad = b.rgradients

    g00 = @inbounds grad[(perm[px0]+perm[py0])&size_mask+1]
    g10 = @inbounds grad[(perm[px1]+perm[py0])&size_mask+1]
    g01 = @inbounds grad[(perm[px0]+perm[py1])&size_mask+1]
    g11 = @inbounds grad[(perm[px1]+perm[py1])&size_mask+1]

    dx = x - x0f
    dy = y - y0f
    dx1 = dx - 1.0
    dy1 = dy - 1.0

    v00 = muladd(g00[1], dx, g00[2] * dy)
    v10 = muladd(g10[1], dx1, g10[2] * dy)
    v01 = muladd(g01[1], dx, g01[2] * dy1)
    v11 = muladd(g11[1], dx1, g11[2] * dy1)

    fx = smooth(dx)
    fy = smooth(dy)

    vx0 = lerp(v00, v10, fx)
    vx1 = lerp(v01, v11, fx)

    return lerp(vx0, vx1, fy)
end

const SYM = [' ', '░', '▒', '▓', '█', '█']

function run(b::Noise, iteration_id::Int64)
    total = UInt32(0)
    size_val = b.size
    y_step = 0.1
    y_start = iteration_id * 12.8

    @inbounds for y = 0:(size_val-1)
        fy = y * y_step + y_start
        @simd for x = 0:(size_val-1)
            fx = x * y_step

            v = (get(b, fx, fy) + 1.0) * 0.5

            idx_f = v * 5.0
            if idx_f >= 5.0
                total += UInt32(SYM[6])
            elseif idx_f <= 0.0
                total += UInt32(SYM[1])
            else
                idx = trunc(Int, idx_f)
                total += UInt32(SYM[idx+1])
            end
        end
    end

    b.result = (b.result + total) & 0xffffffff
end

checksum(b::Noise)::UInt32 = b.result
